File: android/src/main/kotlin/br/com/dextra/bifrost/BifrostCoordinatorChannel.kt
package br.com.dextra.bifrost

import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

internal object BifrostCoordinatorChannel {

  private val channel = MethodChannel(Bifrost.engine.dartExecutor, "bifrost/coordinator")

  init {
    channel.resizeChannelBuffer(2)
  }

  fun onCreatePage(pageData: HashMap<String, Any?>) {
    channel.invokeMethod("onCreatePage", pageData)
  }

  fun onShowPage(pageData: HashMap<String, Any?>) {
    channel.invokeMethod("onShowPage", pageData)
  }

  fun onDeallocPage(pageData: HashMap<String, Any?>) {
    channel.invokeMethod("onDeallocPage", pageData)
  }

  fun onBackPressed(pageData: HashMap<String, Any?>) {
    channel.invokeMethod("onBackPressed", pageData)
  }

  fun canPop(pageData: HashMap<String, Any?>, callback: (Boolean) -> Unit) {
    channel.invokeMethod("canPop", pageData, object : MethodChannel.Result {
      override fun success(result: Any?) = callback((result as? Boolean) ?: false)
      override fun error(code: String?, message: String?, details: Any?) = callback(false)
      override fun notImplemented() = callback(false)
    })
  }

  fun setMethodCallHandler(handler: MethodCallHandler?) {
    channel.setMethodCallHandler(handler)
  }
}

File: ios/Classes/BifrostCoordinatorChannel.swift
//
//  BifrostCoordinatorChannel.swift
//  bifrost
//
//  Created by Matheus Lutero on 23/02/21.
//

import Flutter
import Foundation

internal typealias BifrostPageData = Dictionary<String, Any?>

internal class BifrostCoordinatorChannel {
    
    private var channel: FlutterMethodChannel!
    
    init(_ messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(name: "bifrost/coordinator", binaryMessenger: messenger)
        channel.resizeBuffer(2)
        channel.setMethodCallHandler(handle(_:result:))
    }
    
    func onCreatePage(_ data: BifrostPageData) {
        channel.invokeMethod("onCreatePage", arguments: data)
    }

    func onShowPage(_ data: BifrostPageData) {
        channel.invokeMethod("onShowPage", arguments: data)
    }

    func onDeallocPage(_ data: BifrostPageData) {
        channel.invokeMethod("onDeallocPage", arguments: data)
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "popViewController" {
            Bifrost.popViewController()
        }
    }
}
